### ISSUE TITLE
Discussion: jamstack word treatment

### ISSUE THREAD

**Issue by philhawksworth** [jamstack/jamstack.org#279]
# Discussion: _jamstack_ word treatment

**Note to the reader:**

> This issue discusses how we write "JAMSTACK". I'm going to have to write the word itself a number of times in the issue as a result... which is bound to result in LOLs. So I'll normalise on _jamstack_ for a while. This is not a statement of intent!

## Goals of this issue

- Gather thoughts from the community about the best way to present the word _jamstack_ when written
- Minimize confusion, maximize clarity
- Seek consensus on the best way to avoid confusion when communicating what _jamstack_ is
- It would be useful to be able to be consistent ourselves, and offer a place to find the "generally accepted" word written as a reference.

## Current challenges and observations

- Naming things is hard!
- The word was derived from **J**avaScript, **A**PIs and **M**arkup, but you don't have to use all 3 when adopting this architectureal model. So right away, by writing "JAMstack", we introduce confusion when explaining what this is.
- Most people using the term regularly tends to be using the form "JAMstack"
- The wider community currently use various ways of presenting the word, and we would like to be consistent
- I (Phil 👋) quietly experimented with a revised treatment on jamstack.org without discussing it or announcing it first. This created more confusion and discussion than I anticipated! My bad. Apologies!
- Chris Coyier observered and commented on our flux nicely: https://css-tricks.com/jamstack-vs-jamstack/
- **If** we feel it appropriate to follow the lead of things like "LAMP", then we might note that this is generally referred to as "LAMP stack"
- **If** we would take inspiration from things like "AJAX", then we might note that this is generally referred to as "Ajax" (see below) 
- Naming things is hard!

## Proposal and justification

My opinion is that we should adopt "**Jamstack**" as our norm.

- It avoids the common first question of "what does that stand for?", more often instead prompting, "what is that?"
- It's easier to avoid well-intentioned consistencies than are common with "JAMstack" (JAM Stack, JAM stack, JAMStack)
- and as I mentioned on the _jamstack_ slack:

> I'd just offer this one comparison: Ajax.
>
> When I first started using Ajax, it was always called "AJAX" and stood for Async JavaScript And XML. Over time, JSON became more popular than XML with this pattern, but we still called it the same thing. And these days we'd refer to it as "Ajax" (if at all, lol) because what it stands for is more than what the acronym consists of.

Over the last 5 years since the term was coined, the ecosystem which enables the jamstack has evolved and matured so much. The term needs to not divert attention away from that fact, and towards the original acronym. — I believe that _jamstack_ stands for **so much more** than using JavaScript, APIs and Markup. I think it stands more for:

- Pre-rendering as much as possible
- Serving a site with static assets, without the need for a webserver executing logic to create responses on demand
- Enhancing with JavaScript and leveraging browser APIs where appropriate
- Feasting on the flourishing API economy, and decoupling our own services into APIs which our UIs can consume.

...but I'm keen on wider perspectives about the word.

## For discussion

So... how should we normalise how we write this word?

**Comment by jakeherp** [comment]
I believe JAMstack is most correct since it's an acronym just like MERN stack or LAMP stack vs. Mern and Lamp.

**Comment by philhawksworth** [comment]
That's true @jakeherp. That was its inception.

Any thoughts on when you use this architecture but you are not using APIs or JavaScript? That's where some confusion creeps in.

Ajax started as an acronym too, but the meaning became refined.

**Comment by jakeherp** [comment]
@philhawksworth is it truly a Jamstack site if it isn't using APIs or JS? Just like a React site without Express and MongoDB wouldn't be a MERN site, but simply a React site.
I get your point that a Hugo site is considered Jamstack even though it's based on other technologies, but perhaps that's where the distinction should be different?

**Comment by michrome** [comment]
Did `AJAX` become `Ajax` **because** we send data other than XML, or is that just a coincidence? Are there other examples of acronyms moving from UPPERCASE to Titlecase?

**Comment by philhawksworth** [comment]
> @philhawksworth is it truly a Jamstack site if it isn't using APIs or JS?

Yes it absolutely is.

This is one of the main pieces of confusion that I'm trying to address. Many people are pushing back on jamstack because they believe that it dictates that they MUST use client-side rendering and call out to APIs.

That is a powerful option, and often useful and appropriate. But certainly not a requirement. And often not needed.

The strap-line on jamstack.org is:

> Fast and secure sites and apps delivered by pre-rendering files and serving them directly from a CDN, removing the requirement to manage or run web servers.

And the way I personally elaborate on that is as I did above:

- Pre-rendering as much as possible
- Serving a site with static assets, without the need for a webserver executing logic to create responses on demand
- Enhancing with JavaScript and leveraging browser APIs where appropriate
- Feasting on the flourishing API economy, and decoupling our own services into APIs which our UIs can consume.

**Comment by StarfallProjects** [comment]
@jakeherp I've always understood the "JavaScript" in JAM to be frontend JS, not the tech the static site generator is built with (given you can have SSGs in Go, Python, Java, Bash etc.)

**Comment by michrome** [comment]
> * Pre-rendering as much as possible
> * Serving a site with static assets, without the need for a webserver executing logic to create responses on demand
> * Enhancing with JavaScript and leveraging browser APIs where appropriate
> * Feasting on the flourishing API economy, and decoupling our own services into APIs which our UIs can consume.

Psefstack's got a nice ring to it! 🤡

**Comment by philhawksworth** [comment]
> Did `AJAX` become `Ajax` **because** we send data other than XML, or is that just a coincidence? Are there other examples of acronyms moving from UPPERCASE to Titlecase?

I'm honestly not sure. But we didn't change the Acronym to AJAJ so much as shift towards accepting the term as a description of the model rather than as a strict acronym.

**Comment by StarfallProjects** [comment]
I think one umbrella term for "using a site generator and not worrying about the backend too much" (or Phil's more sophisticated definitions!) is useful. 

I agree JAMStack feels "more correct", and jamstack gets on my tech-writery nerves, but I can see why having to explain "oh it stands for this, but you don't HAVE to use JS or APIs" gets tedious and confusing.

With regard to the AJAX example: while I'd probably write ajax casually, when writing an article/tutorial, or proofreading, I would still correct it to AJAX, as it's an acronym. So perhaps we need to think about what we're deciding on here. Are we discussing how we're going to use the term casually, as a community (in which case, jamstack is inevitable - that shift key is hard to reach!), or if we're going to try and govern a "proper" brand name. On which note - is it an official brand name at all?

I'd suggest that whatever happens, we're stuck with some variation on jam. The term is so widely used. That said, if it ever changes completely, switching to "Marmalade" would be my vote.

**Comment by philhawksworth** [comment]
> That said, if it ever changes completely, switching to "Marmalade" would be my vote.

LOLOL!

_starts designing a logo containing Paddington Bear_

**Comment by philhawksworth** [comment]
> and jamstack gets on my tech-writery nerves...

Just in case there is any confusion or is this was your meaning, I'm advocating for "Jamstack", not "jamstack"

Good points, @StarfallProjects.
I'm not trying to cite Ajax as strict prior art which we should follow to the letter (pun very happily intended)... just think it is an interesting reference.

My hope here is mostly to help us reach agreement on what how we might reasonably expect to see term written consistently on the jamstac.org site, and what conventions we might like to adopt in our wider usage.

jamstack.org is the natural place for people to come to learn "what is this thing?" so at the very least I'd like to have consistent and non-distracting usage here. And like many (I expect) I'd personally adopt what we settle on in my own writings and materials, for the sake of consistency.

**Comment by michrome** [comment]
As @StarfallProjects points out: an English, non-technical reader will interpret a sequence of capital letters as an acronym. Acronyms are useful because:

1. the reader can understand the concept without having to know what the letters stand for. <abbr title="Acquired Immune Deficiency Syndrome">AIDS</abbr> is an example for many people;
2. if the reader does know what the letters stand for, they benefit from that additional information: Acquired Immune Deficiency Syndrome (I had to look it up!).

In contrast, we tend to read words starting with one capital letter as proper nouns e.g. "Netlify", "War and Peace" or "Buckingham Palace". Proper nouns are frequently unambiguous in their definition: a legal entity (Company), a book with an <abbr title="International Standard Book Number">ISBN</abbr> or a specific building in London.

JAMstack seems most align with non-technical English writing in my opinion.

**Comment by colbyfayock** [comment]
i think its a compelling argument that other stacks separate out the word stack from the name, regardless of capitalization. i'm on the JAM camp, but I understand the reasonings here

@philhawksworth do you have any observations from the experiment yet? any "results" i guess?

**Comment by colbyfayock** [comment]
> Just in case there is any confusion or is this was your meaning, I'm advocating for "Jamstack", not "jamstack"

would the logo conform to this? the first version has a lowercase j 🤔

**Comment by n-ham** [comment]
I'm on team JAMstack, even for websites that aren't all made with 'jam'.

**Comment by philhawksworth** [comment]
> JAM stack seems most aligned to non-technical English writing in my opinion.

This is a valuable point, @michrome!  

> I'm personally under the impression that you're going to be led to explaining the situation regardless because people are generally curious. "what's jamstack? oh how did you come up with that?"

Yeah, I think that's true @colbyfayock. Although I currently find myself including 2 extras steps of saying, "It stands for JavaScript, APIs, and Markup"... quickly followed by, "but you don't need to use all 3 of those!" 

I've found that adds a little confusion. And to @michrome's point, if it is an acronym, them making that plain to people without (or with!) English as a first language is useful... as long as the acronym doesn't lead us away from true _meaning_.  

These are all great points, and food for thought!

**Comment by philhawksworth** [comment]
> @philhawksworth do you have any observations from the experiment yet? any "results" i guess?

Only anecdotal. And not at all scientific. I've not been pushing it hard in public channels as I rather wanted to have this conversation first. 

but it seems to land well when I've gone straight to the architectural model, and not put the J, A and M up front.

I suppose a better test is in how people comprehend it when reading about it, rather than being walked through an explanation.

**Comment by remotesynth** [comment]
I am for keeping it JAMstack for now for several reasons.

1. I think the acronym is still useful. Part of why AJAX became Ajax was because the X became less relevant. However, when teaching people what the JAMstack is, the JAM part is still very useful in explaining the parts, even if what it represents is perhaps too vague and the M stands for way too many pieces...but still relevant.
2. I think we're getting too deep into trying to brand this. Afaik, Ajax didn't change because of a branding effort. It just kind of happened. Meanwhile LAMP never became Lamp. We're only about 3 years into introducing the concept and it's only been in the past year that it has really taken hold. That being said, most devs are still new to it or don't even know about it. Shifting branding at that point can cause confusion.

While I agree that the acronym can cause confusion at times, I think we just need to do a better job of clarifying what the acronym is about. I see too many people confused into thinking the J for JavaScript is about JavaScript frameworks - I don't think it is beneficial for the ecosystem for folks to think it requires using React. I see too many people think M is for Markdown rather than markup. Despite these issues, I still find the acronym useful in explaining what the concept is. It adds tangible tools and technologies to an otherwise amorphous concept.

**Comment by chriscoyier** [comment]
I kinda prefer "Jamstack", as this make a lot of sense to me:

> It avoids the common first question of "what does that stand for?", more often instead prompting, "what is that?"

Absolutely non-scientifically, I bet some double-digit % of people think the M stands for Markdown. J and A are like the same thing to me, and I don't really need some architectures permission to use JavaScript to hit an API on my website.  

I just don't think the acronym is all that meaningful and it's more interesting to wonder _what is that?_ and find out.

**Comment by DigitalHodgePodge** [comment]
> I just don't think the acronym is all that meaningful...

While this statement seems pretty accurate seeing as jamstack doesn't _need_ JavaScript or APIs, I do feel that there is some benefit in emphasizing the acronym. One such benefit is the implication that JavaScript and APIs are key technologies for the stack that should be used (when applicable). 

If we are instead reducing their importance in the stack, it seems something like "Flatstack" would be a much more appropriate name. Worse still would be going the MV* route and winding up with **Mstack. :sweat_smile:

All that said, I am mostly playing devil's advocate here, and I kind of like "Jamstack" just for the sake of helping people remain consistent and the ease of typing it out on mobile devices... but I do wonder if it should _officially_ remain "JAMstack" and allow natural progression through usage.

**Comment by philhawksworth** [comment]
Lots of excellent points in this thread! Some which resonate with me more than others, but all very well stated. Thanks everyone!

It looks to me like there is a little more leaning from this group towards "JAMstack" than to "Jamstack" at this point.

I'll leave this discussion open for a while longer and will take all these things into consideration as we work on an iteration to the jamstack.org site.

**Comment by jlengstorf** [comment]
I'll weigh in here. I'm pretty aligned with @chriscoyier on this front:

> I don't really need some architectures permission to use JavaScript to hit an API on my website.

This gets at the core of my issue with "JAM" as an acronym: it’s a meaningless acronym.

MEAN and MERN apps are _also_ using JavaScript, APIs, and Markup — are they Jamstack sites? A PHP site that calls an API and renders markup that includes JavaScript — is that Jamstack?

If we're calling JAM an actual tech stack, these examples use the "stack". but they're definitely _not_ what we mean when we say "Jamstack".

The Jamstack is not a stack; it's a software architecture (like microservices). It's unfortunate that it's phrased as one, but I think we're only making our lives harder if we treat "Jamstack" the same way we treat MEAN or MERN or LAMP, which are _actual_, opinionated stacks.

I think it's impractical to _completely_ rebrand, but I think moving away from an ambiguous acronym and toward a general label that describes an architecture would be a Good Thing™. 💜

**Comment by philhawksworth** [comment]
That's my personal leaning too, @jlengstorf. But I think currently the majority of voices are tending the other way.

We'll keep this thread going a little while longer before we formalise our language on the v2 of the site.

Ultimately, the wider community will do what the wider community does, but we'll at least try to be consistent here :)

**Comment by delster** [comment]
Just wanted to drop by and +1 the `Jamstack` as being clearer to introduce than `JAMstack`, especially in the context of sites/apps that don't check all the `JAM` boxes, where you have to circle around and explain "Jamstack" (little jam) and why it's the same thing.

My head canon is that we started with JAMstack referring to things with JAM; an actual stack, and now we have the Jamstack which has outgrown having JAM as a constraint. 

We could just collectively agree all JAMstacks are Jamstacks but not all Jamstacks are JAMstacks.

**Comment by bnb** [comment]
> The Jamstack is not a stack;

if this is the case then it probably should not include `stack` in the name. 

>I think it's impractical to completely rebrand, but I think moving away from an ambiguous acronym and toward a general label that describes an architecture would be a Good Thing™. 💜

The problem here is that changing from JAM to Jam only serves to introduce further confusion. 

>DEVELOPER: "So it's not a stack?"
>JAMSTACK-Y DEVELOPER: "No, it's an architecture."

This is some galaxy-brain level indirection that amplifies confusion and introduces a further need for explanation of an already confusing concept, which ends up pushing people away rather than bringing them in.

Additionally, any jam (as in the thing you put on bread) branding becomes relatively meaningless since it's just a mush of letters rather than the JAM being a separate entity.

**Comment by bnb** [comment]
For context, how would y'all describe the _architecture_ in 1-3 sentences?

**Comment by philhawksworth** [comment]
Yeah you highlight some clear challenges with the name, @bnb. But we're not talking about changing that name. Especially now, as 5 years after being coined, it is starting to get some real purchase.

Naming stuff is hard. But this is the label we've been using. 

On the site try to summarise concisely with:

> Fast and secure sites and apps delivered by pre-rendering files and serving them directly from a CDN, removing the requirement to manage or run web servers.

**Comment by bnb** [comment]
> But we're not talking about changing that name.

I definitely understand this and where you're coming from, despite fundamentally disagreeing with it ❤️

**Comment by philhawksworth** [comment]
Ok... this discussion has been open for 3 weeks now, and there has been some really helpful conversation. Thanks to all for weighing in!

It was probably a little optimistic to hope that we would all arrive at the exact same conclusion and be 100% decisive and aligned. As different perspectives joined I've personally swayed in both directions. 🙂

***
But let's wrap this up. It looks like just a little more support for those advocating for "**Jamstack**".
***

So, this is the path we will take on the jamstack.org site. And we'll work to do that consistently in the content we add, and in the design.

My hope is, that even if some opinions differ, the site will be consistent and clear, and all the people arriving fresh to the term and the concept, will have a less confused experience. And that it feels more natural to people when writing the word in materials.

Many thanks again.

**Comment by emaildano** [comment]
Great read! it was interesting to see a variety of perspective.  🍻 👍

**Comment by benatkin** [comment]
Good call. Jamstack is the way going forward to bring a wide audience and make it easy to fit into any style guide. JAMstack is our roots.
